### ISSUE TITLE
"Maximum call stack size exceeded" with circular references

### ISSUE THREAD

**Issue by border-radius** [Automattic/mongoose#4046]
Mongoose (4.4.11) falls with error "RangeError: Maximum call stack size exceeded" when `console.log` or `res.json` instance with circular references.
Code to reproduce:
```javascript
var mongoose = require('mongoose');

mongoose.connect('mongodb://localhost/test');

var Foo = mongoose.model('Foo', {
  bar: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Bar'
  }
});

var Bar = mongoose.model('Bar', {
  foo: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Foo'
  }
});

var foo = new Foo();
foo.bar = new Bar({
  foo: foo
});

console.log(foo);
```
Full stack trace:
```
RangeError: Maximum call stack size exceeded
    at model.Document.$toObject (/home/user/playground/node_modules/mongoose/lib/document.js:1882:3)
    at model.Document.toObject (/home/user/playground/node_modules/mongoose/lib/document.js:2089:15)
    at clone (/home/user/playground/node_modules/mongoose/lib/utils.js:252:16)
    at cloneObject (/home/user/playground/node_modules/mongoose/lib/utils.js:300:13)
    at clone (/home/user/playground/node_modules/mongoose/lib/utils.js:258:16)
    at model.Document.$toObject (/home/user/playground/node_modules/mongoose/lib/document.js:1926:13)
    at model.Document.toObject (/home/user/playground/node_modules/mongoose/lib/document.js:2089:15)
    at clone (/home/user/playground/node_modules/mongoose/lib/utils.js:252:16)
    at cloneObject (/home/user/playground/node_modules/mongoose/lib/utils.js:300:13)
    at clone (/home/user/playground/node_modules/mongoose/lib/utils.js:258:16)
    at model.Document.$toObject (/home/user/playground/node_modules/mongoose/lib/document.js:1926:13)
    at model.Document.toObject (/home/user/playground/node_modules/mongoose/lib/document.js:2089:15)
    at clone (/home/user/playground/node_modules/mongoose/lib/utils.js:252:16)
    at cloneObject (/home/user/playground/node_modules/mongoose/lib/utils.js:300:13)
    at clone (/home/user/playground/node_modules/mongoose/lib/utils.js:258:16)
    at model.Document.$toObject (/home/user/playground/node_modules/mongoose/lib/document.js:1926:13)
```

**Comment by vkarpov15** [comment]
Thanks, will look into this tomorrow.

**Comment by vkarpov15** [comment]
I'm gonna mark this as a won't fix: `JSON.stringify()` doesn't work on circular references anyway, the best we can do is crash with a more informative error but that's tricky.

FWIW, in the case you specified, you can use the `depopulate` option to `toJSON()` and `toObject()`: `foo.toJSON({ depopulate: true })`

**Comment by markstos** [comment]
This is a regression. I'm upgrading a code base from Mongoose 3.8.x -> 4.13.x, and I've started getting these crashes due to circular references.

**Comment by vkarpov15** [comment]
@markstos stack trace and code samples please?

**Comment by markstos** [comment]
@vkarpov15 I tested the code posted by the original poster in this ticket. It works for me in Mongoose 3.8.40, and results in "RangeError: Maximum call stack size exceeded" with 4.13.7, just like it does for the OP. My stack trace looks the same as the OPs that has already been shared.  This was tested with Node.Js 4.4.2. 

```
RangeError: Maximum call stack size exceeded
    at model.Document.$toObject (/d/node_modules/mongoose/lib/document.js:2033:3)
    at model.Document.toObject (/d/node_modules/mongoose/lib/document.js:2245:15)
    at clone (/d/node_modules/mongoose/lib/utils.js:254:16)
    at cloneObject (/d/node_modules/mongoose/lib/utils.js:343:13)
    at clone (/d/node_modules/mongoose/lib/utils.js:260:16)
    at model.Document.$toObject (/d/node_modules/mongoose/lib/document.js:2080:13)
    at model.Document.toObject (/d/node_modules/mongoose/lib/document.js:2245:15)
    at clone (/d/node_modules/mongoose/lib/utils.js:254:16)
    at cloneObject (/d/node_modules/mongoose/lib/utils.js:343:13)
    at clone (/d/node_modules/mongoose/lib/utils.js:260:16)
    at model.Document.$toObject (/d/node_modules/mongoose/lib/document.js:2080:13)
    at model.Document.toObject (/d/node_modules/mongoose/lib/document.js:2245:15)
    at clone (/d/node_modules/mongoose/lib/utils.js:254:16)
    at cloneObject (/d/node_modules/mongoose/lib/utils.js:343:13)
    at clone (/d/node_modules/mongoose/lib/utils.js:260:16)
    at model.Document.$toObject (/d/node_modules/mongoose/lib/document.js:2080:13)
```

I also tested the code against `mongoose@5.0.0-rc0` and the regression remains in that version as well. Here's the stack trace for that:

```
/home/mark/tmp/m5/node_modules/mongoose/lib/utils.js:153
exports.clone = function clone(obj, options) {
                              ^

RangeError: Maximum call stack size exceeded
    at clone (/home/mark/tmp/m5/node_modules/mongoose/lib/utils.js:153:31)
    at cloneObject (/home/mark/tmp/m5/node_modules/mongoose/lib/utils.js:246:11)
    at clone (/home/mark/tmp/m5/node_modules/mongoose/lib/utils.js:172:16)
    at model.Document.validate.Document.$toObject (/home/mark/tmp/m5/node_modules/mongoose/lib/document.js:2012:5)
    at model.Document.validate.Document.toObject (/home/mark/tmp/m5/node_modules/mongoose/lib/document.js:2211:15)
    at clone (/home/mark/tmp/m5/node_modules/mongoose/lib/utils.js:166:16)
    at cloneObject (/home/mark/tmp/m5/node_modules/mongoose/lib/utils.js:246:11)
    at clone (/home/mark/tmp/m5/node_modules/mongoose/lib/utils.js:172:16)
    at model.Document.validate.Document.$toObject (/home/mark/tmp/m5/node_modules/mongoose/lib/document.js:2047:13)
    at model.Document.validate.Document.toObject (/home/mark/tmp/m5/node_modules/mongoose/lib/document.js:2211:15)
    at clone (/home/mark/tmp/m5/node_modules/mongoose/lib/utils.js:166:16)
    at cloneObject (/home/mark/tmp/m5/node_modules/mongoose/lib/utils.js:246:11)
    at clone (/home/mark/tmp/m5/node_modules/mongoose/lib/utils.js:172:16)
    at model.Document.validate.Document.$toObject (/home/mark/tmp/m5/node_modules/mongoose/lib/document.js:2047:13)
    at model.Document.validate.Document.toObject (/home/mark/tmp/m5/node_modules/mongoose/lib/document.js:2211:15)
    at clone (/home/mark/tmp/m5/node_modules/mongoose/lib/utils.js:166:16)
```

Is this regression acceptable, or should this ticket be re-opened, since it has a code sample to reproduce the regression?

"call stack exceeded" errors are pretty unpleasant to deal with. Even failing with an error that points to a line of code and recommends using "depopulate" would be preferred over tracking down what's causing this when upgrading Mongoose.

**Comment by vkarpov15** [comment]
I marked this as a won't fix because I'm not certain what the correct behavior should be in the general case of circular references. With OP's example there's an easy workaround, just set `depopulate: true` and everything works, but that's because `depopulate: true` cuts out the circular reference and leaves just an ObjectId. Would `depopulate: true` work in your case @markstos ? I agree that an informative error is better than a call stack error, I just want to make sure it's the right informative error

**Comment by markstos** [comment]
I fixed our app by a combination of using explicit IDs instead of full
objects, and calling `{ depopulate:true }` in some cases.

Because the stack track ends at `toObject()`, I'm not honestly sure what
exactly in our schema started triggering the recursion after we upgraded,
but since I found some workarounds it clearly wasn't essential.

I would say that a hint about trying `depopulate:true` would have been very
helpful in our case.
